### DISCUSSION TITLE
[How To]WSL2+WIN11/WIN10 work good

### DISCUSSION THREAD

**Discussion by Error996** [sickcodes/Docker-OSX#458] [Show and tell]
First you need to make wsl2 work good,for example is ubuntu-20

1. get into wsl,and run these

```
sudo apt install -y tasksel
sudo tasksel install xubuntu-desktop
sudo apt install gtk2-engines
```

2. change to root,vim`~/.bashrc`,add these code

```
export DISPLAY=$(cat /etc/resolv.conf | grep nameserver | awk '{print $2; exit;}'):0.0
export LIBGL_ALWAYS_INDIRECT=1
sudo /etc/init.d/dbus start &> /dev/null
```

3. download `https://sourceforge.net/projects/vcxsrv/` and install

> You’ll be given a choice of four display window options. Let’s choose ‘Multiple windows’ for now and press next.

> On the next screen, you can leave everything as default.

> On the next window, make sure that all options are checked including ‘Disable access control’ then press next, save the configuration if you want and finish.

> You will then probably be asked if you want to allow VcXsrv to access your network, you’ll have to click on the ‘Public network’ option here and allow it through the Windows Firewall or whichever firewall you use.

4. update wsl in powershell

```
wsl --update
wsl --shutdown
```

> open wsl

5. open vcxsrv and read this project's readme

`https://github.com/sickcodes/Docker-OSX#id-like-to-run-docker-osx-on-windows`

6. enjoy it

**Comment by Error996** [comment]
<img width="1839" alt="image" src="https://user-images.githubusercontent.com/12880110/156194875-4dcbfc73-9e77-4b6f-811e-e31d6d8ede14.png">

**Reply by sickcodes** [reply]
Thank you for this!

**Reply by Error996** [reply]
```
To use WSLg's built-in X-11 server, change these two lines in the docker run command to point Docker-OSX to WSLg.

-e "DISPLAY=${DISPLAY:-:0}" \
-v /mnt/wslg/.X11-unix:/tmp/.X11-unix \
```

Change To:

```
To use WSLg's built-in X-11 server, change these two lines in the docker run command to point Docker-OSX to WSLg.

-e "DISPLAY=${DISPLAY:-:0.0}" \
-v /mnt/wslg/.X11-unix:/tmp/.X11-unix \
```

**Comment by MaxRev-Dev** [comment]
Can confirm that PulseAudio is required on Win11+WSL2.
Here's my list of checks:
 - Use pulse audio. This should fix audio driver error https://github.com/sickcodes/Docker-OSX/issues/91
 - Check distro path for WSLg, e.g it can be `/mnt/ubuntu22` on Ubuntu 22.04
 - Check if GUI is enabled. REMOVE `guiApplications=false` from `.wslconfig`
 - KVM is enabled by default, but you can add `nestedVirtualization=true` to `.wslconfig`
 - No need to install VcXsrv when using WSLg.
 
```bash
docker run    \
 --device /dev/kvm  \
 -e AUDIO_DRIVER=pa,server=unix:/tmp/pulseaudio.socket \
 -v /mnt/ubuntu22/wslg/runtime-dir/pulse/native:/tmp/pulseaudio.socket   \
 -v /mnt/ubuntu22/wslg/.X11-unix:/tmp/.X11-unix  \
  sickcodes/docker-osx:monterey
```

**Reply by softlion** [reply]
I keep having error:

pulseaudio: pa_context_connect() failed
pulseaudio: Reason: Connection refused
pulseaudio: Failed to initialize PA contextaudio: Could not init `pa' audio driver

Maybe you know the solution ?

**Reply by softlion** [reply]
Fixed the issue ! use /mnt/wslg instead of ubuntu22 (even in ubuntu22 !)

```
 sudo apt install pulseaudio -y
 pulseaudio --version

 docker run     --device /dev/kvm   -e AUDIO_DRIVER=pa,server=unix:/tmp/pulseaudio.socket  -v /mnt/wslg/runtime-dir/pulse/native:/tmp/pulseaudio.socket    -v /mnt/wslg/.X11-unix:/tmp/.X11-unix  -e PULSE_SERVER=unix:/tmp/pulseaudio.socket    sickcodes/docker-osx:monterey
```

**Reply by MaxRev-Dev** [reply]
I use several distros, so mine was in one of the distro paths.
For newcomers:
- Double check the folder of WLSg within your distro folder or `/mnt`.

**Reply by softlion** [reply]
Well the issue was, the path you set also exists ! But the command would fail.

**Comment by alexpatcas** [comment]
I would pay  money for someone to help me get this working on windows 11

**Comment by bennakarim2306** [comment]
Hey all,

thank you for the support provided by some responses.
I am stuck to a point, where the installation needs ~=2gb but i have only 1gb allocated for it, as you can see here:
![image](https://github.com/sickcodes/Docker-OSX/assets/39828744/8aaeaf97-f5ca-4d9a-9799-1bc92d094d7b)

Some context:

- Using WSL2 on windows 11
- Command to run docker: `docker run     --device /dev/kvm   -e AUDIO_DRIVER=pa,server=unix:/tmp/pulseaudio.socket  -v /mnt/wslg/runtime-dir/pulse/native:/tmp/pulseaudio.socket    -v /mnt/wslg/.X11-unix:/tmp/.X11-unix  -e PULSE_SERVER=unix:/tmp/pulseaudio.socket    sickcodes/docker-osx:monterey`

I am pretty sure, i either need to set-up a config param by running the docker container or allocating more space to the wsl.

By running the command "df" in ubuntu, i got following: 
![image](https://github.com/sickcodes/Docker-OSX/assets/39828744/f918c30f-e083-4c09-aefc-2239b11ddff3)

so i guess 8158368 (1K-Blocks ~= 3.8GB) should be enough. Please correct me, if i am saying non-sense here.

Thank you for the support and KR

**Reply by bennakarim2306** [reply]
I found the solution: using mac os Catalina and not Monterey, as explained here: https://github.com/sickcodes/Docker-OSX?tab=readme-ov-file#run-catalina-pre-installed-

**Comment by ioweb-gr** [comment]
I've tried the instructions with tahoe and I assigned 32G RAM and 12 cores to the WSL2 config. I also followed through with the install of the system but I notice crazy lag when trying to move the cursor or select options to the point that it's almost unusable. 

```
export VERSION=tahoe
docker run -it \
  --name docker-osx-$VERSION \
  --device /dev/kvm \
  -p 50922:10022 \
  -e DISPLAY=:0 \
  -e RAM=28 \
  -e GENERATE_UNIQUE=false \
  -e CPU=Haswell-noTSX \
  -e CPUID_FLAGS=kvm=on,vendor=GenuineIntel,+invtsc,vmware-cpuid-freq=on \
  -e MASTER_PLIST_URL=https://raw.githubusercontent.com/sickcodes/osx-serial-generator/master/config-custom-sonoma.plist \
  -e SHORTNAME=$VERSION \
  -v /mnt/wslg/.X11-unix:/tmp/.X11-unix \
  -v docker-osx_data:/home/arch/OSX-KVM \
  sickcodes/docker-osx:latest
```

https://github.com/user-attachments/assets/e3e176d6-d353-45bb-a58c-8941f410dcc9

Any idea what I can do to make it perform better?
